### PR TITLE
feat: add DynamoDB Cache and Queue drivers

### PR DESCRIPTION
**Pull request by kcwinner**
feat: add in dynamodb cache driver
feat: add in dynamodb queue driver

**Check List**
- [x] Tests has been run in packages where changes made if available
- [x] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Description of Changes Made (if issue reference is not provided)**

* Added a new QueryCacheDriver and QueryQueueDriver to support DynamoDB
  * Add new environment variable options:
    1. `CUBEJS_CACHE_TABLE`
    1. `DYNAMODB_TTL_KEY`
  * Adds new `dynamodb` option for `CUBEJS_CACHE_AND_QUEUE_DRIVER` environment variable
* Fixes a small annoyance with the Aurora Serverless types file

### CODE REVIEW

File: packages/cubejs-query-orchestrator/src/orchestrator/QueryCache.js [hunk 1]
```diff
@@ -12,9 +13,23 @@ export class QueryCache {
...
+        this.cacheDriver = new RedisCacheDriver({ pool: options.redisPool });
+        break;
+      case 'dynamodb':
+        this.cacheDriver = new DynamoDBCacheDriver({ tableName: options.tableName });
+        break;
+      case 'memory':
+        this.cacheDriver = new LocalCacheDriver();
+        break;
+      default:
+        throw new Error('Cache Not Specified')
```
Comment (kcwinner) [inline packages/cubejs-query-orchestrator/src/orchestrator/QueryCache.js:31]: Should this default to memory? It gets passed in from QueryOrchestrator so it should never actually be empty

**Comment by hassankhan** [comment]
@kcwinner Love the PR!

I was wondering if you'd had a chance to look at https://github.com/cube-js/cube.js/issues/663? When we last discussed the issue, [we thought](https://github.com/cube-js/cube.js/issues/663#issuecomment-636367515) that we'd need to use [DynamoDB Transactions](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transactions.html). Curious to know how well it works without 😀

**Comment by kcwinner** [comment]
@hassankhan Ahhhh thanks for sharing. I did not see this before. The cache portion is "simple". Wasn't aware of the other use just from reading through

Currently works fine when using locally (with run dev and playground). Working to get it working in a Lambda function setup, but struggling with the package structure since it's not released in npm and server-core relies on it. Once I get that piece working it should give me a better idea about transactions.

**Comment by paveltiunov** [comment]
@kcwinner Hey Kenneth! Thanks for contributing this! So this is a cache driver however it lacks implementation of a queue driver. So Redis would still be required. Was it your initial thought on how do you want to use it?

**Comment by kcwinner** [comment]
@paveltiunov I realized that after seeing @hassankhan's message and investigating a bit further. Also explains why everything works when using the playground / local dev mode and not when to deployed to AWS Lambda.

I'm working on the DynamoDB queue now. My goal is to run serverless and without the vpc requirement. I'll push the queue once it's a bit cleaner

I investigated lambda.store for use but quickly hit the free limit (5k requests). Not sure the pricing model is ideal for use with cubejs given the cost per request and egress pricing.

**Comment by kcwinner** [comment]
It's *mostly* working now. Still some buggy issues - mostly things I think I'm just not familiar with as far as cubejs goes.

I occasionally run into the following error. Sometimes the data pulls fine, sometimes it returns
`Error: TypeError: Cannot read property 'map' of undefined`
I've tried tracing from the api-gateway but I must be missing something.

```
Performing query completed: 50db1441-ae3d-473b-823e-a3e3f2b1d758-span-3 (2397ms)
Slow Query Warning: 50db1441-ae3d-473b-823e-a3e3f2b1d758-span-4 
--
"SELECT `orders__completed_at_day` `orders__completed_at_day`, sum(`orders__count`) `orders__count` FROM stb_pre_aggregations.orders_some_rollup GROUP BY 1 ORDER BY 1 ASC LIMIT 10000"
--
{} 
Query is too slow to be renewed during the user request and was served from the cache. Please consider using low latency pre-aggregations.
Internal Server Error: 50db1441-ae3d-473b-823e-a3e3f2b1d758-span-4 (5416ms)
--
{
  "measures": [
    "Orders.count"
  ],
  "timeDimensions": [
    {
      "dimension": "Orders.completedAt",
      "granularity": "day"
    }
  ],
  "order": {},
  "filters": [],
  "dimensions": []
}
--
{
  "authInfo": {
    "iat": 1607217733,
    "exp": 1607304133
  }
} 
TypeError: Cannot read property 'map' of undefined
    at transformData (/Users/ken/github/cube-js/cube.js/packages/cubejs-server-core/node_modules/@cubejs-backend/api-gateway/index.js:112:91)
    at /Users/ken/github/cube-js/cube.js/packages/cubejs-server-core/node_modules/@cubejs-backend/api-gateway/index.js:620:17
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 0)
    at ApiGateway.load (/Users/ken/github/cube-js/cube.js/packages/cubejs-server-core/node_modules/@cubejs-backend/api-gateway/index.js:593:23)
    at /Users/ken/github/cube-js/cube.js/packages/cubejs-server-core/node_modules/@cubejs-backend/api-gateway/index.js:374:7
Performing query completed: 50db1441-ae3d-473b-823e-a3e3f2b1d758-span-3 (8064ms)
```

**Comment by kcwinner** [comment]
Sorry for the delay in getting back to this. I've got it working now (locally). Next step is to find the best way to bundle it and deploy to my lambda function. 

Hoping to have it mostly cleaned up by end of year.

File: packages/cubejs-query-orchestrator/src/orchestrator/DynamoDBCacheDriver.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    });
+  }
+
+  public async remove(key: string) {
+    await this.cache.delete({ pk: CACHE_TABLE_PRIMARY_KEY, key });
+  }
+
+  public async keysStartingWith(prefix) {
+    // TODO: validate this works
+    console.log('### KEYS STARTING WITH RESULT');
```
Comment (hassankhan) [inline packages/cubejs-query-orchestrator/src/orchestrator/DynamoDBCacheDriver.ts:70]: Might want to remove this before the PR is merged 😄

Comment (kcwinner) [reply packages/cubejs-query-orchestrator/src/orchestrator/DynamoDBCacheDriver.ts]: Removed!

**Comment by kcwinner** [comment]
Tests are still finicky. I could use some feedback if someone is able/willing to run them that is more familiar with cube. Otherwise I am running it internally at my company. There may be some bugs and things missing but they haven't come to light heavily since I'm the only one using it

`yarn build && yarn run integration:dynamodb`
```
QueryQueueDynamoDB
    ✓ gutter (1702 ms)
    ✓ instant double wait resolve (1450 ms)
    ✕ priority (1304 ms)
    ✓ timeout (2278 ms)
    ✕ stage reporting (85 ms)
    ✕ priority stage reporting (144 ms)
    ✕ negative priority (1487 ms)
    ✕ orphaned (5706 ms)
    ✓ removed before reconciled (1317 ms)
    ✓ queue driver lock obtain race condition (280 ms)
    ✓ activated but lock is not acquired (329 ms)
```

**Comment by hobojoe** [comment]
Looking forward to this! Not needing redis would be very nice!

**Comment by paveltiunov** [comment]
@kcwinner Hey Kenneth! Thanks for putting a lot of effort into this one! Are these tests have intermittent fails or always failing? This test suite ensures consistency guarantees for the query queue. Basically, if those guarantees aren't achieved queue may end up in an undefined state. We'd need to ensure DynamoDB can provide those guarantees before we move on with a merge. Do you think you have some additional time to work towards fixing those? Thanks!

**Comment by kcwinner** [comment]
@paveltiunov Many of the tests seem to fail due to the implicit timing in the tests. I believe the DynamoDB driver will be a bit slower than redis by nature. One of the main issues is just lack of fully understanding the tests / timing holistically on my end. I will work to take another dive here soon. The biggest offenders were the stage / priority based ones

**Comment by hobojoe** [comment]
Having DynamoDB as a Cache option would certainly reduce cloud costs on small projects!

**Comment by moltar** [comment]
Would be amazing to have this working! I'm no DynamoDB expert, but here are a few ideas.

- Use DynamoDB local for testing faster (https://hub.docker.com/r/amazon/dynamodb-local/)
- Use AWS SDK v3 package family, which also has many Dynamo convenience packages

**Comment by hobojoe** [comment]
1 year later ... still waiting for dynamodb as cache!

**Comment by igorlukanin** [comment]
I'd like to thank everyone who invested time and effort into this PR. However, given the [move from Reds to Cube Store](https://cube.dev/blog/replacing-redis-with-cube-store) as a custom-tailored solution for cache and queue in Cube and a subsequent [removal of Redis support](https://cube.dev/blog/how-you-win-by-using-cube-store-part-1#how-to-migrate-to-cube-store), finally [completed in version 0.36](https://cube.dev/blog/cube-core-v0-36-playground-2-0-custom-granularities-query-pushdown#breaking-removed-features), I feel like this PR is very unlikely to move further. So let me close it but feel free to reopen if anything.

**Pull request closed without merging**